### PR TITLE
Use Map

### PR DESCRIPTION
**Pull request by na-yuri**
- Updated `chessGame` to have map structure
- Checked in `move()` if there is an en passant in legalMoves and if there is and it was not chosen to delete it from legalMoves (now thinking that we don't even need this since `updateLegalMoves()` is called before every move is made)
- Need to change first move after first move (for King and Rook)
- Moved Rook right away if a King castling move was made
- Updated `Box` to have an operator< (this is used for iterating through a map)

### CODE REVIEW

File: piece.cc [hunk 1]
```diff
@@ -39,11 +39,12 @@ int Piece::move(Piece *currentTile, Piece *targetTile, int newX, int newY) {
...
     */
 
-    updateLegalMoves();
+    std::map<Box, int> tempLegalMovesMap = updateLegalMoves();
+    legalMovesMap = &tempLegalMovesMap;
 
     const Box targetBox(newX, newY);
 
-    if (std::find(legalMovesArr->begin(),legalMovesArr->end(), targetBox) != legalMovesArr->end()) {
+    if (legalMovesMap->find(targetBox) != legalMovesMap->end()) {
```
Comment (rogerz74) [inline piece.cc:47]: do u need to add the `std::` to the `find()` function here?

File: piece.cc [hunk 2]
```diff
@@ -52,10 +53,80 @@ int Piece::move(Piece *currentTile, Piece *targetTile, int newX, int newY) {
...
+        if (targetTile->getName() == "K" && (*legalMovesMap)[targetBox] == 2) {
+            // to find which rook we are moving towards
+            if (yCoord - newY > 0) {
+                for (auto &pair: (*((*board)[7][0]->getLegalMoves()))) {
+                    if (pair.second == 2) {
+                        std::swap((*board)[7][0], (*board)[pair.first.getX()][pair.first.getY()]);
+                        (*board)[pair.first.getX()][pair.first.getY()]->xCoord = 7;
+                        (*board)[pair.first.getX()][pair.first.getY()]->yCoord = 0;
+                        (*board)[7][0]->xCoord = pair.first.getX();
+                        (*board)[7][0]->yCoord = pair.first.getY();
```
Comment (rogerz74) [inline piece.cc:66]: B/c the `(*board)[pair.first.getX()][pair.first.getY()]` is now pointing at the rook AFTER THE SWAP, shouldn't the `xCoord` and `yCoord` be just updated to `pair.first.getX()` and `pair.first.getY()`?

Also, B/c the rook is moving into a nullptr, taking `(*board)[7][0]->xCoord` will result in an error b/c `*board)[7][0]` is null. Hence, you don't even need to change any xCoords or yCoords for nullptrs on the board.

File: piece.cc [hunk 3]
```diff
@@ -52,10 +53,80 @@ int Piece::move(Piece *currentTile, Piece *targetTile, int newX, int newY) {
...
+                for (auto &pair: (*((*board)[7][7]->getLegalMoves()))) {
+                    if (pair.second == 2) {
+                        std::swap((*board)[7][7], (*board)[pair.first.getX()][pair.first.getY()]);
+                        (*board)[pair.first.getX()][pair.first.getY()]->xCoord = 7;
+                        (*board)[pair.first.getX()][pair.first.getY()]->yCoord = 7;
+                        (*board)[7][7]->xCoord = pair.first.getX();
+                        (*board)[7][7]->yCoord = pair.first.getY();
+                    }
+                }
+            }
```
Comment (rogerz74) [inline piece.cc:79]: Same comment above applies here and also for the black king castle below

File: piece.cc [hunk 4]
```diff
@@ -39,11 +39,12 @@ int Piece::move(Piece *currentTile, Piece *targetTile, int newX, int newY) {
 
     */
 
-    updateLegalMoves();
+    std::map<Box, int> tempLegalMovesMap = updateLegalMoves();
+    legalMovesMap = &tempLegalMovesMap;
```
Comment (rogerz74) [inline piece.cc:43]: Does this work? I thought this was the thing that was giving you trouble lol.

File: piece.cc (hunk 1 above)
Comment (na-yuri) [reply piece.cc:47]: No because this find isn't in the standard library it is a function for maps

File: piece.cc (hunk 4 above)
Comment (na-yuri) [reply piece.cc:43]: idk apparently this works I wasn't able to do
`legalMovesMap = &updateLegalMoves()`
also apparently map has an operator= and a copy constructor

File: piece.cc (hunk 2 above)
Comment (na-yuri) [reply piece.cc:66]: you are totally right

**Review by rogerz74 — APPROVED** [review]

**Pull request merged**